### PR TITLE
page_service: add peer_addr span field, and set tenant_id / timeline_id fields earlier

### PR DESCRIPTION
**Pull request by problame**
Before this PR, during shutdown, we'd find naked logs like this one for every active page service connection:

```
2023-07-05T14:13:50.791992Z  INFO shutdown request received in run_message_loop
```

This PR
1. adds a peer_addr span field to distinguish the connections in logs
2. sets the tenant_id / timeline_id fields earlier

It would be nice to have `tenant_id` and `timeline_id` directly on the `page_service_conn_main` span (empty, initially), then set them at the top of `process_query`.

But, the problem is that the debug asserts for tenant_id and timeline_id presence in the tracing span doesn't support detecting empty values (https://github.com/neondatabase/neon/issues/4676)
So, I'm a bit hesitant about over-using `Span::record`.

### CODE REVIEW

**Review by conradludgate — APPROVED** [review]

File: pageserver/src/page_service.rs [hunk 1]
```diff
@@ -531,7 +539,7 @@ impl PageServerHandler {
         Ok(())
     }
 
-    #[instrument(skip(self, pgb, ctx))]
+    #[instrument(skip_all, fields(start_lsn=%start_lsn, end_lsn=%end_lsn))]
```
Comment (koivunej) [inline pageserver/src/page_service.rs:542]:
```suggestion
    #[instrument(skip_all, fields(%start_lsn, %end_lsn))]
```

File: pageserver/src/page_service.rs [hunk 2]
```diff
@@ -738,7 +747,7 @@ impl PageServerHandler {
     }
 
     #[allow(clippy::too_many_arguments)]
-    #[instrument(skip(self, pgb, ctx))]
+    #[instrument(skip_all, fields(lsn=?lsn, prev_lsn=?prev_lsn, full_backup=%full_backup))]
```
Comment (koivunej) [inline pageserver/src/page_service.rs:750]:
```suggestion
    #[instrument(skip_all, fields(?lsn, ?prev_lsn, %full_backup))]
```

**Review by koivunej — APPROVED** [review]
Good changes, would minimize the field decls away.

**Comment by problame** [comment]
Applied your suggestions.

Before I merge, wdyt regarding:

> I'm wondering whether I can have tenant_id and timeline_id directly on the page_service_conn_main span (empty, initially), then set them at the top of process_query?

Some more context. With this patch, the log lines look like this:

```
2023-07-05T17:06:55.632061Z  INFO page_service_conn_main{peer_addr=127.0.0.1:52706}:process_query{tenant_id=95144dc32ceea9a11bb86f4d0c3ab092 timeline_id=5359ccaff96843b105dca762b63dfc5b}:handle_basebackup_request{lsn=None prev_lsn=None full_backup=false}: taking basebackup lsn=0/16960E8, prev_lsn=0/1696070 (full_backup=false)
```

I'd prefer the tenant and timeline id were in the `page_service_conn_main` span:

```
2023-07-05T17:06:55.632061Z  INFO page_service_conn_main{peer_addr=127.0.0.1:52706 tenant_id=95144dc32ceea9a11bb86f4d0c3ab092 timeline_id=5359ccaff96843b105dca762b63dfc5b}:process_query:handle_basebackup_request{lsn=None prev_lsn=None full_backup=false}: taking basebackup lsn=0/16960E8, prev_lsn=0/1696070 (full_backup=false)
```

We could even leave out the `process_query` span

```
2023-07-05T17:06:55.632061Z  INFO page_service_conn_main{peer_addr=127.0.0.1:52706 tenant_id=95144dc32ceea9a11bb86f4d0c3ab092 timeline_id=5359ccaff96843b105dca762b63dfc5b}:process_query:handle_basebackup_request{lsn=None prev_lsn=None full_backup=false}: taking basebackup lsn=0/16960E8, prev_lsn=0/1696070 (full_backup=false)
```

It doesn't make strict sense at the protocol level though, because, tenant and timeline id are provided per request. 

Any thoughts?

**Comment by koivunej** [comment]
No objections to:

> We could even leave out the `process_query` span [and record tenant_id and timeline_id to the `page_service_conn_main`]

Should be quite trivial to do:
- just remove the `#[instrument(...)]` around `process_query`
- then Span::record will target the `page_service_conn_main`

Just need to make sure we have the `debug_assert_tenant_and_timeline_ids_...` (or tenant_id only) however that was in all query handlers, in case something is changed on the pg-connection-handling-in-anonymous-inner-class abstraction.

**Comment by problame** [comment]
> Just need to make sure we have the debug_assert_tenant_and_timeline_ids_... (or tenant_id only) however that was in all query handlers, in case something is changed on the pg-connection-handling-in-anonymous-inner-class abstraction. These seem to be already in place.

Yeah, that's what I'm worried about.

Also the `debug_assert_tenant_and...()`: do they still trap if there's a field, but it's the empty `tracing::Value`?

Gonna play with this for 10min or so to find out.

**Comment by problame** [comment]
```rust

    #[test]
    #[should_panic]
    fn empty_field_defined_in_span_macro_panics() {
        let setup = setup_current_thread();

        let span = tracing::info_span!("root", tenant_id = tracing::field::Empty);
        let _guard = span.enter();
        check_fields_present([&setup.tenant_extractor]).unwrap();
    }

    #[test]
    #[should_panic]
    fn empty_field_defined_by_instrument_macro_panics() {
        let setup = setup_current_thread();

        #[tracing::instrument(skip_all, fields(tenant_id))]
        fn child_fn(setup: &Setup) {
            check_fields_present([&setup.tenant_extractor]).unwrap();
        }

        let span = tracing::info_span!("root", tenant_id = tracing::field::Empty);
        let _guard = span.enter();
        child_fn(&setup);
    }
```

Adding these tests, they fail (`libs/utils/src/tracing_span_assert.rs`).

Thinking about it a bit more, even if we detect the empty value correctly, there's no way to assert that the value is not stale.

I.e., imagine

```
page_service_conn_main
  process_query
    it's a basebackup request for tenant_id=A, timeline_id=B
    sets tenant_id=A, timeline_id=B
  process_query
    it's a getpage request for tenant_id=C, timeline_id=D
    we forgot the Span::record() calls for this branch
    we'll log in the name of `tenant_id=A, timeline_id=B`
  
```

What is worse:
* like it is now, with `peer_addr` in `page_service_conn_main` and `tenant_id` and `timeline_id` in the `process_query` span, where staleness risk isn't big because each process_query gets a new `empty` value
* having all 3 in `page_service_conn_main` span, but the ids may be stale if there's a bug

**Comment by problame** [comment]
Hm, the current state of this PR is an improvement, and it has near-zero chance of the staleness issue outlined above.

Merging it as is.

---

Regarding my observations about the debug assert not triggering on `tracing::value::Empty`. 

I'll create a follow-up issue and report there why it's not easily fixable.

**Pull request merged**